compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2605,9 +2605,30 @@ pub enum Rvalue<'tcx> {
 static_assert_size!(Rvalue<'_>, 40);
 
 impl<'tcx> Rvalue<'tcx> {
+    /// Returns true if evaluating rvalue can have side effects.
     #[inline]
-    pub fn is_pointer_int_cast(&self) -> bool {
-        matches!(self, Rvalue::Cast(CastKind::PointerAddress, _, _))
+    pub fn may_have_side_effects(&self) -> bool {
+        match self {
+            // Pointer to int casts may be side-effects due to exposing the provenance.
+            // While the model is undecided, we should be conservative. See
+            // <https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html>
+            Rvalue::Cast(CastKind::PointerAddress, _, _) => true,
+
+            Rvalue::Use(_)
+            | Rvalue::Repeat(_, _)
+            | Rvalue::Ref(_, _, _)
+            | Rvalue::ThreadLocalRef(_)
+            | Rvalue::AddressOf(_, _)
+            | Rvalue::Len(_)
+            | Rvalue::Cast(CastKind::Misc | CastKind::Pointer(_), _, _)
```

Comment:
Int-to-ptr casts should be considered as making a non-deterministic choice (angelically) and reading mutable machine state (namely the set of exposed provenances), which can prevent them from being reordered with other operations.

Though I assume reads from mutable machine state are fine here since otherwise `Use` would also be out entirely. So my guess is the property you care about here is "can this rvalue be removed if we don't care about the result of the computation"?